app/lib/error.py

```diff
@@ -1,45 +1,252 @@
+        if self.data:
+            app.logger.error('[{}][DATA]: {}'.format(self.error_type,
+                                                     self.data))
+
+    def _form_message(self, message, subs):
+        """Apply subsitutions to the error message if any exist.  If there are
+        spots for subsitutions variables in the error message (if the strings
+        ``'%s'`` appears anywhere in the string) and there are not enough
+        strings in ``subs`` to fill them, instances of ``'%s'`` will be removed
+        from the error message.  If there are two many strings in ``subs``,
+        later strings will be ignored.
+
+        Here are some examples::
+
+            >>> _form_message('Error at url `%s`: %s', ['/home', 'Not Found'])
+            'Error at url `/home`: Not Found'
+            >>> _form_message('Error at url `%s`: %s', ['/home'])
+            'Error at url `/home`: '
+            >>> _form_message('Error at url.', ['/home'])
+            'Error at url.'
+
+        :param str message: The format string to format.
+        :param subs: The strings to subsitute into the format string.
+        :type subs: list(str)
+
+        :returns: The formatted string.
+        :rtype: str
+        """
+        n_subs = message.count('%s')
+
+        # Extend subs if there aren't enough
+        if n_subs > len(subs):
+            subs.extend([''] * (n_subs - len(subs)))
+
+        # Shorten subs if there are too many
+        if len(subs) > n_subs:
+            subs = subs[:n_subs]
+
+        # Perform substitutions
+        if subs:
+            message = message % subs
+
+        return re.sub(r'\%s', '', message)
+
+
+def _make_subclasses(error_data, baseclass):
+    """Recursively add subclass errors to ``baseclass``.
+
+    :param dict error_data: A dictionary of error name strings to a four-tuple
+        of ``(message, error_code, http_status_code, subclasses)``. The
+        ``message`` is the user-facing message to display, the ``error_code``
+        is a unique number representing the error, the ``http_status_code`` is
+        the HTTP status code that should be associated with this error should
+        it surface to the user, and subcalsses is a dictionary of this same
+        type, containing data for subclasses of this error.
+    :param baseclass: The error class to attach these errors onto.
+    :type baseclass: :class:`EventumError` or one of its subclasses.
+    """
+    if not error_data:
+        return
+
+    for e_type, e_attrs in error_data.iteritems():
+        # Make the new class object
+        name = '{}.{}'.format(baseclass.__name__, e_type)
+        subclass = classobj(name, (baseclass,), {})
+
+        # Configure it's message, error_code, and http_status_code
+        subclass.message = e_attrs[0]
+        subclass.error_code = e_attrs[1]
+        subclass.http_status_code = e_attrs[2]
+
+        # Recursively generate subclasses
+        _make_subclasses(e_attrs[3], subclass)
```

Comment:
i think it would be cleaner if you unpacked everything in `e_attrs` at once and then used those temp vars afterwards